massa-factory-worker/src/denunciation_factory.rs

```diff
@@ -0,0 +1,303 @@
+                factory.run();
+            })
+            .expect("failed to spawn thread : denunciation-factory")
+    }
+
+    /// Process new secured header (~ block header)
+    fn process_new_secured_header(&mut self, secured_header: SecuredHeader) {
+        let key = secured_header.content.slot;
+
+        if self.block_header_by_slot.len() > DENUNCIATION_FACTORY_BLOCK_HEADER_CACHE_MAX_LEN {
+            warn!(
+                "Denunciation factory cannot process - cache full: {}",
+                self.block_header_by_slot.len()
+            );
+            return;
+        }
+
+        // Get selected address from selector and check
+        // Note: If the public key of the header creator is not checked to match the PoS,
+        //       someone can spam with headers coming from various non-PoS-drawn pubkeys
+        //       and cause a problem
+        let selected_address = self.channels.selector.get_producer(key);
+        match selected_address {
+            Ok(address) => {
+                if address != Address::from_public_key(&secured_header.content_creator_pub_key) {
+                    warn!("Denunciation factory received a secured header but address was not selected");
+                    return;
+                }
+            }
+            Err(e) => {
+                warn!("Cannot get producer from selector: {}", e);
+            }
+        }
+
+        let denunciation_: Option<Denunciation> = match self.block_header_by_slot.entry(key) {
+            Entry::Occupied(mut eo) => {
+                match eo.get_mut() {
+                    BlockHeaderDenunciationStatus::Accumulating(header_1_) => {
+                        let header_1: &SecuredHeader = header_1_;
+                        match Denunciation::try_from((header_1, &secured_header)) {
+                            Ok(de) => {
+                                eo.insert(BlockHeaderDenunciationStatus::DenunciationEmitted);
+                                Some(de)
+                            }
+                            Err(e) => {
+                                debug!("Denunciation factory cannot create denunciation from endorsements: {}", e);
+                                None
+                            }
+                        }
+                    }
+                    BlockHeaderDenunciationStatus::DenunciationEmitted => {
+                        // Already 2 entries - so a Denunciation has already been created
+                        None
+                    }
+                }
+            }
+            Entry::Vacant(ev) => {
+                ev.insert(BlockHeaderDenunciationStatus::Accumulating(secured_header));
+                None
+            }
+        };
+
+        if let Some(denunciation) = denunciation_ {
+            info!(
+                "Created a new block header denunciation : {:?}",
+                denunciation
+            );
+
+            self.channels.pool.add_denunciation(denunciation);
+        }
+
```

Comment:
if think it can be simplified to this 

```suggestion
match self.block_header_by_slot.entry(key) {
    Entry::Occupied(mut eo) => {
        match eo.get_mut() {
            BlockHeaderDenunciationStatus::Accumulating(header_1_) => {
                let header_1: &SecuredHeader = header_1_;
                match Denunciation::try_from((header_1, &secured_header)) {
                    Ok(de) => {
                        eo.insert(BlockHeaderDenunciationStatus::DenunciationEmitted);
                        info!("Created a new block header denunciation : {:?}", de);
                        self.channels.pool.add_denunciation(de);
                    }
                    Err(e) => {
                        debug!("Denunciation factory cannot create denunciation from endorsements: {}", e);
                    }
                }
            }
            // Denunciation has already been created
            BlockHeaderDenunciationStatus::DenunciationEmitted => {
            }
        }
    }
    Entry::Vacant(ev) => {
        ev.insert(BlockHeaderDenunciationStatus::Accumulating(secured_header));
    }
};

```

Reply:
Not a big fan to have the process of denunciation creation and the storage all in one. I'd rather keep my way, clearly seeing the steps...